### PR TITLE
Autoregressive mode and embedding calculation addition

### PR DESCRIPTION
**Pull request by metric-space**
## Changes
1. Add in `autoregressive` flag to notify `AutoModelSequenceEmbeddings` and `AutoModelForRagE2E` that an autoregressive model is being used 
2. Add autoregressive model of computation of embeddings based on last hidden state
3. Move logic around so `mean_pooling` is more applicable to both scenarios (`clm` and `mlm`) 
4. update cli args
5. pad_token = eos_token if autorgessive
6. add is_autoregressive flag to eval portion
7. fix if to elif in save hook
8. remove default save statement
9. unwrap model properly while saving

### CODE REVIEW

File: dalm/models/retriever_only_base_model.py [hunk 1]
```diff
@@ -30,18 +31,25 @@ def __init__(
...
-        model_output = self.model(input_ids, attention_mask)
-        embeddings = self.mean_pooling(model_output, attention_mask)
+        if self.is_autoregressive:
+            # we take the last hidden state of the model
+            token_embeddings = self.model(
+                input_ids, attention_mask=attention_mask, output_hidden_states=True, return_dict=True
+            ).hidden_states[-1]
+        else:
+            # First element of model_output contains all token embeddings
+            token_embeddings = self.model(input_ids, attention_mask)[0]
```
[the remark this answers is not in the document]
Comment (metric-space) [reply dalm/models/retriever_only_base_model.py:45]: I think that is existing code

File: dalm/models/retriever_only_base_model.py [hunk 2]
```diff
@@ -30,18 +31,25 @@ def __init__(
...
+        embeddings = self.mean_pooling(token_embeddings, attention_mask)
```
[the remark this answers is not in the document]
Comment (metric-space) [reply dalm/models/retriever_only_base_model.py:46]: I believe the shape of the input token embeddings is `(1,<length of tokenized inputs>, <number of features>)`

The output of hidden states flag is of shape [number of layers, 1, `length of tokenized inputs`, `number_of_features`]

so I think we're doing number 2 here?

[the remark this answers is not in the document]
Comment (shamanez) [reply dalm/models/retriever_only_base_model.py:46]: Got it. I guess we can do the same trick for encoder-only models as well.

**Comment by shamanez** [comment]
I guess we can remove the me embedding addition thing .. with a note we can say we got better results by taking the eos token as the representation ..

**Review by shamanez — APPROVED** [review]
Looks good

**Comment by Serega6678** [comment]
BGE models require CLS pooling and not the mean pooling

https://huggingface.co/BAAI/bge-large-en#frequently-asked-questions
